include/ibamr/AdvDiffHierarchyIntegrator.h

```diff
@@ -474,6 +487,21 @@ class AdvDiffHierarchyIntegrator : public IBTK::HierarchyIntegrator
      */
     int getResetPriority(SAMRAI::tbox::Pointer<SAMRAI::pdat::CellVariable<NDIM, double> > Q_var) const;
 
+    /*!
+     * \brief Register BrinkmanPenalizationAdvDiff object to add Brinkman penalization terms to the advection-diffusion
```

Comment:
The way that we handle this for IB methods for FSI is to have the IB solver tell the field solver about forces through a `CartGridFunction`. That way, the Navier-Stokes solver is oblivious to the details of the IB formulation.

Is it possible to use the same design pattern here?

Reply:
Possibly? I'm not entirely sure. This method requires masking the diffusion coefficient and the convective operator, so the solvers need to be adjusted prior to each cycle.

Reply:
In that case, would it make sense for this to be a subclass of `AdvDiffSemiImplicitHierarchyIntegrator`?

Reply:
I think that would be the best option.

Reply:
It looks like the methods that would be needed in the subclass are mostly already provided by the Brinkmann class, and so it shouldn't be much work to move the code from `AdvDiffSemiImplicitHierarchyIntegrator` to the Brinkmann class.

@Raamkrishnan, is it clear what is being suggested and does it make sense? Before doing anything --- @drwells, do you have an opinion?

Reply:
`CartGridFunction` are not useful, unless everything could be described analytically, which never happens in practice and stuff related to level sets. One needs to do some (complex) computations to arrive at a forcing field. This computation is usually integrated with time-stepping scheme as well.

Reply:
I disagree with the statement that `CartGridFunction` is not useful. :-)

When I was looking at the interface changes, I did not realize that the solver was being changed under the hood, and so the explicit coupling through forcing functions is not what you want to do.

Reply:
BrinkmanPenalizationAdvDiff is a helper object that helps modify solver coefficients, when penalization is desired. It is not an integrator, hence we did not subclass from AdvDiffSemiImplicitHI.

Reply:
> I disagree with the statement that `CartGridFunction` is not useful. :-)

(The complex fluid solver is completely built on top of `CartGridFunction`)

Reply:
So is our turbulence code. But we can not use them at *all*  places.

Reply:
> I disagree with the statement that `CartGridFunction` is not useful. :-)
> 
> When I was looking at the interface changes, I did not realize that the solver was being changed under the hood, and so the explicit coupling through forcing functions is not what you want to do.

Yes, we are modifying solver coefficients, masking convection and other forces in the fictitious region, and in case of flux BCs, we are computing a forcing term that in general is a complex function of level set and other scalars. Moreover, there could be some interfaces that have Dirichlet BC and some may have Neumann BCs; all of this taken care by this helper class.